class Solution {
    void solve( int[] nums, List<List<Integer>> ans, List<Integer> output, int index){
        if(index==nums.length){
            ans.add(new ArrayList<>(output));
            return;
        }
        output.add(nums[index]);
        solve(nums, ans, output, index+1);
        output.remove(output.size()-1);
        solve(nums, ans, output, index+1);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> output=new ArrayList<>();
        int index=0;
        solve( nums, ans, output, index );
        return ans;


        
    }
}